dev-extbase-fluid: Some fixes to OAI-PMH interface

dvoracek-slub opened the pull request:
This fixes some issues that I noticed while writing tests (#756).

---

- Fix listing records and identifiers (set $resultSet['elements'])
- Fix XML error when embedding METS documents (there has been an excess angle bracket)
- Don't include <resumptionToken> tag when there's no token
- Fix value of cursor attribute in <resumptionToken>
- Don't show request parameters when there's a badVerb or badArgument
- Restore fallback "earliestDatestamp" when no document is found

Commit:
Various fixes in OAI-PMH interface

- Fix listing records and identifiers (set $resultSet['elements'])
- Fix XML error when embedding METS documents (there has been an excess
  angle bracket)
- Don't include <resumptionToken> tag when there's no token
- Fix value of cursor attribute in <resumptionToken>
- Don't show request parameters when there's a badVerb or badArgument
- Restore fallback "earliestDatestamp" when no document is found

albig commented on Classes/Controller/OaiPmhController.php:
```diff
@@ -240,10 +240,7 @@ protected function getMetsData(array $record)
-                $mets = $xml->saveXML();
-                // Remove leading line
-                $mets = substr($mets, strpos($mets, '>'));
+                $mets = $xml->saveXML($root->item(0));
```
so easy. Thank you :-)

dvoracek-slub replied:
:-)

albig commented on Classes/Controller/OaiPmhController.php:
```diff
@@ -764,7 +776,7 @@ protected function generateResumptionTokenForDocumentListSet(array $documentList
         $resumptionTokenInfo = [];
         $resumptionTokenInfo['token'] = $resumptionToken;
-        $resumptionTokenInfo['cursor'] = $documentListSet['metadata']['completeListSize'] - count($documentListSet);
+        $resumptionTokenInfo['cursor'] = $currentCursor;
```
This is correct for the `cursor` attribute following the linked docu. But we should change the XSLT file for the HTML-Output for this, don't we?

![Kitodo Presentation  OAI 2 0 Request Results](https://user-images.githubusercontent.com/3354500/154691310-8c821073-c084-4920-b236-f04663e7588f.png)

Before your change, there was written "10" (for 10 documents per request).

dvoracek-slub replied:
I'm not sure what's expected here, but if it's just a more user-friendly representation of the XML file, perhaps we could just clarify that it's the "*Previously* Submitted Records"?

albig replied:
Unfortunately, we do not have the amount of items in the current list. I've just looked around, how other solved this.

albig commented on Resources/Private/Templates/OaiPmh/Main.xml:
```diff
@@ -122,7 +124,9 @@
                     </f:if>
                 </f:for>
-                <resumptionToken cursor="{resumptionToken.cursor}" completeListSize="{resumptionToken.completeListSize}" expirationDate="<f:format.date format="Y-m-d\TH:i:s\Z">{resumptionToken.expired}</f:format.date>">{resumptionToken.token}</resumptionToken>
+                <f:if condition="{resumptionToken.token}">
```
The `resumptionToken` is missing now on the last "page".

In the linked "FlowControl" it's written:

> The repository responds with an incomplete list of 75 records. The repository marks this list as the final incomplete list by including in the response an empty resumptionToken element with two attributes: a completeListSize of 175, and a cursor of 100.

dvoracek-slub replied:
Sorry for having missed that. Do you think the patch is right? As far as I understood, the `<resumptionToken>` tag should still be omitted if there is only one page, and there shouldn't be an `expirationDate` parameter on the last page. So I've extracted this to a separate section in the Fluid template.

albig reviewed (changes requested):
Thank you for working on that. Please have a look at my comments.

Commit:
Fix using "set" parameter in OAI-PMH

Commit:
Fix handling of OAI-PMH resumption token

- The last batch of an incomplete list contains an empty resumptionToken
- If the first batch is already complete, there is no resumptionToken tag

Commit:
Tweak resumptionToken tag. (#1)

* Always output the resumptionToken tag.

This adds again the resumptionToken if we have a complete list without a
token itself. This is not a must feature but quite common and I think
it's not forbidden.

This changes the XSLT stylesheet a bit because we output the "cursor" in
the bottom which is not the amount of submitted records but the position
in the list of records.

Just another extample of a complete list:

https://www.tib.eu/oai/public/repository/open?verb=ListIdentifiers&metadataPrefix=oai_dc&set=zbmql

* Revert "Always output the resumptionToken tag."

Convinced. I should return an empty resumptionToken tag even if it is
not forbidden.

This reverts commit 079265aa1a187da119c439cd70ea705bc3458ae5.

* Change missleading label

Commit:
Merge branch 'dev-extbase-fluid' into dev-oaipmh-fixes
